gyp/pylib/gyp/generator/msvs.py

```diff
@@ -2651,9 +2653,17 @@ def _GetMSBuildGlobalProperties(spec, guid, gyp_file_name):
       ]
     ]
 
+  preferred_tool_architecture = None
+
   if os.environ.get('PROCESSOR_ARCHITECTURE') == 'AMD64' or \
      os.environ.get('PROCESSOR_ARCHITEW6432') == 'AMD64':
-    properties[0].append(['PreferredToolArchitecture', 'x64'])
+     preferred_tool_architecture = 'x64'
+
+  if os.environ.get('npm_config_arch') == 'arm64':
+    preferred_tool_architecture = 'ARM64'
```

Comment:
This sets the PreferredToolArchitecture property in the vcxproj files. This property is [used to select either the x64 or x86 tool exe files to use during the build](https://docs.microsoft.com/en-us/cpp/build/msbuild-visual-cpp-overview?view=vs-2017#preferredtoolarchitecture-property).

For example, in a standard VS2017 Community installation with the ARM64 tools installed, there are six versions of cl.exe:

```cmd
c:\Program Files (x86)\Microsoft Visual Studio\2017\Community\vc\Tools\MSVC\14.16.27023\bin\Hostx64\arm64\cl.exe
c:\Program Files (x86)\Microsoft Visual Studio\2017\Community\vc\Tools\MSVC\14.16.27023\bin\Hostx64\x64\cl.exe
c:\Program Files (x86)\Microsoft Visual Studio\2017\Community\vc\Tools\MSVC\14.16.27023\bin\Hostx64\x86\cl.exe
c:\Program Files (x86)\Microsoft Visual Studio\2017\Community\vc\Tools\MSVC\14.16.27023\bin\Hostx86\arm64\cl.exe
c:\Program Files (x86)\Microsoft Visual Studio\2017\Community\vc\Tools\MSVC\14.16.27023\bin\Hostx86\x64\cl.exe
c:\Program Files (x86)\Microsoft Visual Studio\2017\Community\vc\Tools\MSVC\14.16.27023\bin\Hostx86\x86\cl.exe
```

This property sets the 'Hostx64' part of the path. Since there is no 'HostARM64' folder--that is, Visual Studio 2017 does not have a native ARM64 compiler--setting PreferredToolArchitecture to ARM64 is incorrect. When cross-compiling from x86 or x64 to ARM64, either the x86 or x64 tools should be used.

Note that if the host is ARM64, only the x86 tools will run (emulation). So [if `PROCESSOR_ARCHITEW6432` is set to ARM64](https://docs.microsoft.com/en-us/windows/desktop/WinProg64/wow64-implementation-details) then PreferredToolArchitecture should be x86.

Reply:
Sorry for the delayed response. Thanks for clarifying this. I misunderstood that 'PreferredToolArchitecture' is. I will fix this based on your suggestion.

Reply:
Great to hear. You're taking on cross compiling in Windows, and I really appreciate it.

Reply:
Fixed, based on your comments.